wp-content/plugins/tapestry/controller/class.tapestry-controller.php

```diff
@@ -3,4 +3,106 @@
  * Add/update/retrieve Tapestry post and its child nodes
  * 
  */
-class TapestryController {}
\ No newline at end of file
+
+class TapestryController {
+    const POST_TYPES = [
+        'TAPESTRY' => 'tapestry',
+        'TAPESTRY_NODE' => 'tapestry_node'
+    ];
+    const ERRORS = [
+        'INVALID_POST_ID' => [
+            'MESSAGE' => 'PostID is invalid',
+            'STATUS' => ['status' => 404]
+        ]
+    ];
+    private $postId;
+
+    /**
+     * Constructor
+     */
+    public function __construct($postId) {
+        $this->postId = $postId;
+    }
+
+    /**
+     * Update Tapestry nodes first then
+     * Update the existing Tapestry if the postId is provided
+     * Otherwise, a new Tapestry will be created
+     * 
+     * @param  Object @tapestry The Tapestry data
+     * @return Object @tapestry
+     */
+    public function updateTapestry($tapestry) {
+        // TODO: use isValidPostID() utlility function
+        if (!isset($this->postId)) {
+            $this->postId = $this->_updatePost($tapestry, 'tapestry');
+        }
+
+        $this->_updateNodes($tapestry->nodes, $this->postId);
+
+        if (!isset($tapestry->rootId)) {
+            $tapestry->rootId = $tapestry->nodes[0]->id;
+        }
+
+        $tapestry->nodes = $this->_getNodeIds($tapestry->nodes);
+
+        update_post_meta($this->postId, 'tapestry', $tapestry);
+        return $tapestry;
+    }
+
+    private function _updateNodes($nodes, $postId) {
+        foreach ($nodes as $node) {
+            if (!isset($node->id)) {
+                $nodePostId = $this->_updatePost($node, 'tapestry_node');
+                $metadata = $this->_makeMetadata($node, $nodePostId);
+                $node->id = add_post_meta($postId, 'tapestry_node', $metadata);
+            } else {
+                $metadata = get_metadata_by_mid('post', $node->id)->meta_value;
+                $nodePostId = $metadata->post_id;
+            }
+            update_post_meta($nodePostId, 'tapestry_node_data', $node);
+        }
+    }
+
+    private function _updatePost($post, $postType = 'tapestry', $postId = null) {
+        switch($postType) {
```

Comment:
I would suggest that if post type is `tapestry` and the class property `postId` is set already, then you should set it here so it saves it rather than create a new tapestry.